### PR TITLE
Display synced protocol drugs

### PR DESCRIPTION
**Pull request by saket**

### CODE REVIEW

File: app/src/main/java/org/simple/clinic/facility/Facility.kt [hunk 1]
```diff
@@ -33,6 +34,8 @@ data class Facility(
 
     val pinCode: String?,
 
+    val protocolUuid: UUID? = null,
```
Comment (saket) [inline app/src/main/java/org/simple/clinic/facility/Facility.kt]: I think we should also mention in a comment why this is nullable.

File: app/src/androidTest/java/org/simple/clinic/di/TestAppComponent.kt [hunk 2]
```diff
@@ -19,6 +19,7 @@ import org.simple.clinic.overdue.communication.CommunicationRepositoryAndroidTes
 import org.simple.clinic.overdue.communication.CommunicationSyncAndroidTest
```
Comment (saket) [inline app/src/androidTest/java/org/simple/clinic/di/TestAppComponent.kt]: Let's rename this commit to just `Add method to fetch drugs for a protocol`. It's implied that the work will also be accompanied with automated tests.

File: app/src/androidTest/java/org/simple/clinic/protocolv2/ProtocolRepositoryAndroidTest.kt [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+  @get:Rule
+  val authenticationRule = AuthenticationRule()
+
+  @Before
+  fun setUp() {
+    TestClinicApp.appComponent().inject(this)
+  }
+
+  // TODO: Remove user auth when this test is removed.
+  // TODO: Make protocolUuid non-null in both Facility and FacilityPayload.
```
Comment (saket) [inline app/src/androidTest/java/org/simple/clinic/protocolv2/ProtocolRepositoryAndroidTest.kt]: This TODO can be updated to `Make protocolUuid non-null in FacilityPayload.`

File: app/src/androidTest/java/org/simple/clinic/protocolv2/ProtocolRepositoryAndroidTest.kt [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+  fun when_protocols_are_not_present_in_database_then_default_drugs_should_be_returned() {
+    database.clearAllTables()
+
+    val randomProtocolId = UUID.randomUUID()
+    val drugs = protocolRepository.drugsForProtocol(randomProtocolId).blockingFirst()
+    assertThat(drugs).isEqualTo(protocolRepository.defaultProtocolDrugs())
+  }
+
+  @Test
+  fun drugs_are_present_but_feature_is_disabled_then_default_drugs_should_be_returned() {
```
Comment (saket) [inline app/src/androidTest/java/org/simple/clinic/protocolv2/ProtocolRepositoryAndroidTest.kt]: Looks like we missed `when` in the test name

File: app/src/androidTest/java/org/simple/clinic/protocolv2/ProtocolRepositoryAndroidTest.kt [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    val config = configProvider.blockingGet()
+    if (config.isProtocolDrugSyncEnabled.not()) {
+      val protocol1 = testData.protocol()
+      database.protocolDao().save(listOf(protocol1))
+
+      val drug1 = testData.protocolDrug(protocolUuid = protocol1.uuid)
```
Comment (saket) [inline app/src/androidTest/java/org/simple/clinic/protocolv2/ProtocolRepositoryAndroidTest.kt]: `drug1` -> `drug`

File: app/src/androidTest/java/org/simple/clinic/protocolv2/ProtocolRepositoryAndroidTest.kt [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    val randomProtocolId = UUID.randomUUID()
+    val drugs = protocolRepository.drugsForProtocol(randomProtocolId).blockingFirst()
+    assertThat(drugs).isEqualTo(protocolRepository.defaultProtocolDrugs())
+  }
+
+  @Test
+  fun drugs_are_present_but_feature_is_disabled_then_default_drugs_should_be_returned() {
+    val config = configProvider.blockingGet()
+    if (config.isProtocolDrugSyncEnabled.not()) {
+      val protocol1 = testData.protocol()
```
Comment (saket) [inline app/src/androidTest/java/org/simple/clinic/protocolv2/ProtocolRepositoryAndroidTest.kt]: `protocol1` -> `protocol`

File: app/src/androidTest/java/org/simple/clinic/protocolv2/ProtocolRepositoryAndroidTest.kt [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+      database.protocolDao().save(listOf(protocol1))
+
+      val drug1 = testData.protocolDrug(protocolUuid = protocol1.uuid)
+      database.protocolDrugDao().save(listOf(drug1))
+
+      val drugs = protocolRepository.drugsForProtocol(protocol1.uuid).blockingFirst()
```
Comment (saket) [inline app/src/androidTest/java/org/simple/clinic/protocolv2/ProtocolRepositoryAndroidTest.kt]: `drugs` -> `fetchedDrugs`?

File: app/src/androidTest/java/org/simple/clinic/protocolv2/ProtocolRepositoryAndroidTest.kt [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+      assertThat(drugs).isEqualTo(protocolRepository.defaultProtocolDrugs())
+    }
+  }
+
+  @Test
+  fun when_drugs_are_not_present_for_a_specific_protocol_then_default_values_should_be_returned() {
```
Comment (saket) [inline app/src/androidTest/java/org/simple/clinic/protocolv2/ProtocolRepositoryAndroidTest.kt]: `specific` could be removed

File: app/src/androidTest/java/org/simple/clinic/protocolv2/ProtocolRepositoryAndroidTest.kt [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+
+    val drugsForProtocol2 = protocolRepository.drugsForProtocol(protocol2.uuid).blockingFirst()
+    assertThat(drugsForProtocol2).isEqualTo(protocolRepository.defaultProtocolDrugs())
+  }
+
+  @Test
+  fun when_protocols_are_present_in_database_then_they_should_be_returned() {
+    val config = configProvider.blockingGet()
+
+    if (config.isProtocolDrugSyncEnabled) {
```
Comment (saket) [inline app/src/androidTest/java/org/simple/clinic/protocolv2/ProtocolRepositoryAndroidTest.kt]: Prefer using early returns to avoid indenting a large code block. This can be changed to,

```kotlin
if (config.isProtocolDrugSyncEnabled.not()) {
  return
}
```

File: app/src/main/java/org/simple/clinic/protocolv2/ProtocolModule.kt [hunk 10]
```diff
@@ -31,6 +31,9 @@ class ProtocolModule {
     return retrofit.create(ProtocolSyncApiV2::class.java)
   }
 
+  @Provides
+  fun provideProtocolDrugSyncEnabledConfig(): Single<ProtocolConfig> = Single.just(ProtocolConfig(true))
```
Comment (saket) [inline app/src/main/java/org/simple/clinic/protocolv2/ProtocolModule.kt]: I think we should keep it disabled in this commit

File: app/src/androidTest/java/org/simple/clinic/TestData.kt [hunk 11]
```diff
@@ -602,7 +602,7 @@ class TestData @Inject constructor(
       uuid: UUID = UUID.randomUUID(),
```
Comment (saket) [inline app/src/androidTest/java/org/simple/clinic/TestData.kt]: `repository` is getting truncated in this commit message. Let's remove `from repository` entirely? It's already obvious that all data come from repositories.

File: app/src/androidTest/java/org/simple/clinic/drugs/PrescriptionRepositoryAndroidTest.kt [hunk 12]
```diff
@@ -56,44 +55,20 @@ class PrescriptionRepositoryAndroidTest {
...
+    val amlodipine5mg = testData.protocolDrug(name = "Amlodipine", dosage = "5mg", protocolUuid = protocolUuid)
 
-    repository.savePrescription(patientUuid, amlodipine, "5mg").blockingAwait()
+    repository.savePrescription(patientUuid, amlodipine5mg).blockingAwait()
 
     val savedPrescriptions1 = repository.newestPrescriptionsForPatient(patientUuid).blockingFirst()
     assertThat(savedPrescriptions1).hasSize(1)
 
-    repository.savePrescription(patientUuid, amlodipine, "10mg")
+    repository.savePrescription(patientUuid, amlodipine5mg)
```
Comment (saket) [inline app/src/androidTest/java/org/simple/clinic/drugs/PrescriptionRepositoryAndroidTest.kt]: This looks like a mistake. We should be saving 10mg here. Correct code:

```kotlin
val protocolUuid = UUID.randomUUID()
val amlodipine5mg = testData.protocolDrug(name = "Amlodipine", dosage = "5mg", protocolUuid = protocolUuid)
val amlodipine10mg = testData.protocolDrug(name = "Amlodipine", dosage = "10mg", protocolUuid = protocolUuid)

repository.savePrescription(patientUuid, amlodipine5mg).blockingAwait()

val savedPrescriptions = repository.newestPrescriptionsForPatient(patientUuid).blockingFirst()
assertThat(savedPrescriptions).hasSize(1)

repository.savePrescription(patientUuid, amlodipine10mg)
    .andThen(repository.softDeletePrescription(savedPrescriptions.first().uuid))
    .blockingAwait()

val savedPrescriptionsAfterDelete = repository.newestPrescriptionsForPatient(patientUuid).blockingFirst()
assertThat(savedPrescriptionsAfterDelete).hasSize(1)
```

File: app/src/androidTest/java/org/simple/clinic/MigrationAndroidTest.kt [hunk 13]
```diff
@@ -711,6 +708,41 @@ class MigrationAndroidTest {
...
+        'Country',
+        'Pin code',
+        '2018-09-25T11:20:42.008Z',
+        '2018-09-25T11:20:42.008Z',
+        'PENDING',
+        '2018-09-25T11:20:42.008Z')
+    """)
+
+    val db_v25 = helper.migrateTo(25, Migration_24_25())
+    db_v25.assertColumnCount(tableName = "Facility", expectedCount = 14)
```
Comment (sanchita) [inline app/src/androidTest/java/org/simple/clinic/MigrationAndroidTest.kt]: We can add the assertions in one place i.e next to line# `738`

File: app/src/main/java/org/simple/clinic/protocolv2/ProtocolRepository.kt [hunk 14]
```diff
@@ -59,10 +65,99 @@ class ProtocolRepository @Inject constructor(
...
+  }
+
+  @VisibleForTesting
+  fun defaultProtocolDrugs(): List<ProtocolDrugAndDosages> {
+    val protocolDrug = { uuid: UUID, name: String, dosage: String ->
+      ProtocolDrug(
+          uuid = uuid,
+          protocolUuid = UUID.fromString("d486b15b-ac2a-4d2f-b87c-e1905d5ee791"),
+          name = name,
+          rxNormCode = "rxNormCode-1",
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/protocolv2/ProtocolRepository.kt]: We really should be getting the actual rx norm codes instead of using dummy ones.

File: app/src/main/java/org/simple/clinic/protocolv2/ProtocolRepository.kt [hunk 15]
```diff
@@ -59,10 +65,99 @@ class ProtocolRepository @Inject constructor(
...
+
+    val defaultDrugs = protocolDrugsWithConfig
+        .filter { (drugs, config) -> drugs.isEmpty() || config.isProtocolDrugSyncEnabled.not() }
+        .map { defaultProtocolDrugs() }
+
+    return syncedDrugs.mergeWith(defaultDrugs)
+  }
+
+  @VisibleForTesting
+  fun defaultProtocolDrugs(): List<ProtocolDrugAndDosages> {
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/protocolv2/ProtocolRepository.kt]: I'm wondering if we should inject the default drugs as well. This will then restrict the hardcoding of drugs to outside this repository.

Comment (saket) [reply app/src/main/java/org/simple/clinic/protocolv2/ProtocolRepository.kt]: I considered doing this as well but it was too late. I'd now do this improvement later if that's okay

Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/protocolv2/ProtocolRepository.kt]: Yeah, not necessary to do it in this PR.

File: app/src/androidTest/java/org/simple/clinic/protocolv2/ProtocolRepositoryAndroidTest.kt [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+
+  @Before
+  fun setUp() {
+    TestClinicApp.appComponent().inject(this)
+  }
+
+  // TODO: Remove user auth when this test is removed.
+  // TODO: Make protocolUuid non-null in both Facility and FacilityPayload.
+  @Test
+  fun protocol_ID_should_not_be_empty_when_feature_is_enabled() {
```
[the remark this answers is not in the document]
Comment (saket) [reply app/src/androidTest/java/org/simple/clinic/protocolv2/ProtocolRepositoryAndroidTest.kt]: Agreed

[the remark this answers is not in the document]
Comment (saket) [reply app/src/androidTest/java/org/simple/clinic/protocolv2/ProtocolRepositoryAndroidTest.kt]: On a second thought, I'm going to replace this with another test that ensures `FacilityPayload#protocolUuid` is not null once the feature is enabled.

```kotlin
@Test
fun protocol_ID_in_payload_should_not_be_empty_when_feature_is_enabled() {
  val config = configProvider.blockingGet()
  if (config.isProtocolDrugSyncEnabled) {
    val facilities = facilitySyncApi.pull(recordsToPull = 100, lastPullToken = null).blockingGet()
    val facilityWithoutProtocolId = facilities.payloads.find { it.protocolUuid == null }
    assertThat(facilityWithoutProtocolId).isNull()
  }
}
```

File: app/src/main/java/org/simple/clinic/protocolv2/ProtocolRepository.kt (hunk 14 above)
Comment (saket) [reply app/src/main/java/org/simple/clinic/protocolv2/ProtocolRepository.kt]: I will do this

File: app/src/test/java/org/simple/clinic/drugs/selection/PrescribedDrugsScreenControllerTest.kt [hunk 17]
```diff
@@ -37,100 +43,104 @@ class PrescribedDrugsScreenControllerTest {
...
     val prescriptionUuid1 = UUID.randomUUID()
     val prescriptionUuid2 = UUID.randomUUID()
     val prescriptionUuid3 = UUID.randomUUID()
     val prescriptionUuid4 = UUID.randomUUID()
 
-    val amlodipinePrescription = PatientMocker.prescription(name = "Amlodipine", dosage = "10mg", isProtocolDrug = true)
-    val telmisartanPrescription = PatientMocker.prescription(uuid = prescriptionUuid1, name = "Telmisartan", dosage = "9000mg", isProtocolDrug = false)
+    val amlodipine10mgPrescription = PatientMocker.prescription(name = "Amlodipine", dosage = "10mg", isProtocolDrug = true)
+    val telmisartan9000mgPrescription = PatientMocker.prescription(uuid = prescriptionUuid1, name = "Telmisartan", dosage = "9000mg", isProtocolDrug = false)
     val ReesesPrescription = PatientMocker.prescription(uuid = prescriptionUuid2, name = "Reese's", dosage = "5 packets", isProtocolDrug = false)
```
Comment (vinaysshenoy) [inline app/src/test/java/org/simple/clinic/drugs/selection/PrescribedDrugsScreenControllerTest.kt]: Capital letter? 😅

Comment (saket) [reply app/src/test/java/org/simple/clinic/drugs/selection/PrescribedDrugsScreenControllerTest.kt]: I'll fix this

Comment (vinaysshenoy) [reply app/src/test/java/org/simple/clinic/drugs/selection/PrescribedDrugsScreenControllerTest.kt]: Assuming this will be fixed in another PR. Resolving this comment now.
